### DISCUSSION TITLE
next/font/local: Allow selective preload of sources

### DISCUSSION THREAD

**Discussion by darthmaim** [vercel/next.js#46699] [Ideas]
### Goals

* Allow to configure preload of fonts per file, instead of per font

### Background

Currently either all or none of the files for a font are preloaded. Sometimes it makes sense to only preload some variants, and only load others when needed (for example italics or rarely used weights).

This is especially useful when using variables to style `<b>` or `<i>` elements.

The current workaround for this would look like this:

```ts
const font = localFont({
  src: [
    { path: './font-400.woff2', weight: '400' },
    { path: './font-700.woff2', weight: '700' },
  ],
  variable: '--font',
});

const fontOther = localFont({
  src: [
    { path: './font-400-italic.woff2', weight: '400', style: 'italic' },
    { path: './font-700-italic.woff2', weight: '700', style: 'italic' },
  ],
  variable: '--font-other',
  preload: false
});
```
```css
body { font-family: var(--font); }
i {
  font-style: italic;
  font-family: var(--font-other);
}
```

It is required to define the font-family for the `<i>`-tag to use the correct font. But this can cause problems when the `<i>` is wrapped in a parent with a different font-family, which might have an italic variant already.

### Proposal

A possible solution is to define the font like this:

```ts
const font = localFont({
  src: [
    { path: './font-400.woff2', weight: '400' },
    { path: './font-700.woff2', weight: '700' },
    { path: './font-400-italic.woff2', weight: '400', style: 'italic', preload: false },
    { path: './font-700-italic.woff2', weight: '700', style: 'italic', preload: false },
  ],
  variable: '--font',
});
```

This will only preload the commonly used variants, without the need to change the font-family for the others and avoiding the inheritance problem described above.

This is a small change and I have a proof of concept PR ready: https://github.com/vercel/next.js/compare/canary...darthmaim:next.js:feature/font-local-src-preload

**Comment by mismosmi** [comment]
This would be really cool. We are seeing suboptimal lighthouse scores because of this.

**Comment by MangoMarcus** [comment]
+1 for this

**Comment by larsschellhas** [comment]
I also came across this issue as Firefox is complaining about a lot of preloaded fonts which aren't being used in the first seconds.

There also seems to be another PR targeting this issue here: https://github.com/vercel/next.js/pull/69900

**Comment by indeox** [comment]
If this helps anyone, turns out there is a way to do this now, even though it's not immediately obvious.

The _"[declarations](https://nextjs.org/docs/app/api-reference/components/font#declarations)"_ option can set `font-family`, and when it does, the loader skips the generated name it would assign otherwise.

So two calls can declare the *same* family and carry *different* `preload` flags. The browser sees one family with all the faces; and only the one you split out gets a preload tag.

```ts
// fonts.ts
import localFont from "next/font/local";

// The face that blocks first paint — preloaded on its own.
export const MyFont = localFont({
  variable: "--my-font",
  preload: true,
  declarations: [{ prop: "font-family", value: "MyFont" }],
  src: [
    { path: "./myfont-regular.woff2", weight: "400", style: "normal" }
  ],
});

// Everything else — same family, not preloaded.
export const MyFontDeferred = localFont({
  variable: "--my-font-deferred",
  preload: false,
  adjustFontFallback: false,
  declarations: [{ prop: "font-family", value: "MyFont" }],
  src: [
    { path: "./myfont-italic.woff2", weight: "400", style: "italic" },
    { path: "./myfont-bold.woff2", weight: "700", style: "normal" },
  ],
});
```

```tsx
// layout.tsx — apply both
<html className={`${MyFont.variable} ${MyFontDeferred.variable}`}>
```

You get one `@font-face` group under `MyFont`, so weight/style matching resolves normally, and a single `<link rel="preload">` for the 400. 

It still would be nice to have per-source `preload`. But this works for now.

_This is on Next v16.2.11 with Turbopack_

**Reply by stefanprobst** [reply]
thanks for this - for the css variable to be correct, the `font-family` value in `declarations` must be the same as the exported binding it seems.
